src/api/dynamic_type.c

```diff
@@ -188,6 +188,114 @@ rosidl_dynamic_typesupport_dynamic_type_builder_clone(
 }
 
 
+static rcutils_ret_t
+handle_nested_type(
+  rosidl_dynamic_typesupport_serialization_support_t * serialization_support,
+  const rosidl_runtime_c__type_description__TypeDescription * description,
+  rcutils_allocator_t * allocator,
+  rosidl_dynamic_typesupport_dynamic_type_builder_t * dynamic_type_builder,
+  rosidl_runtime_c__type_description__Field * field,
+  size_t field_offset)
+{
+  if (field->type.nested_type_name.data == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Nested type name from nested field [%s]", field->name.data);
+    return RCUTILS_RET_ERROR;
+  }
+
+  // Create a new type description to pass to the next layer
+  rosidl_runtime_c__type_description__TypeDescription * recurse_desc = NULL;
+  rosidl_runtime_c__type_description__IndividualTypeDescription * recurse_indiv_desc = NULL;
+
+  // NOTE(methylDragon): recurse_indiv_desc borrows from
+  //                     description->referenced_type_descriptions.
+  //                     It is NOT a copy!! Do NOT finalize, modify, or delete it!
+  rcutils_ret_t ret = rosidl_runtime_c_type_description_utils_find_referenced_type_description(
+    &description->referenced_type_descriptions,
+    field->type.nested_type_name.data,
+    &recurse_indiv_desc);
+  if (ret != RCUTILS_RET_OK || recurse_indiv_desc == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Could not find referenced type description [%s]", field->type.nested_type_name.data);
+    return ret;
+  }
+
+  ret = rosidl_runtime_c_type_description_utils_get_referenced_type_description_as_type_description(
+    &description->referenced_type_descriptions, recurse_indiv_desc, &recurse_desc,
+    true);  // Coerce to valid
+  if (ret != RCUTILS_RET_OK || recurse_desc == NULL) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Could not get referenced type description as full type description [%s]",
+      field->type.nested_type_name.data);
+    if (ret == RCUTILS_RET_OK) {
+      ret = RCUTILS_RET_ERROR;
+    }
+    return ret;
+  }
+
+  // Recurse
+  rosidl_dynamic_typesupport_dynamic_type_builder_t nested_type_builder =
+    rosidl_dynamic_typesupport_get_zero_initialized_dynamic_type_builder();
+  nested_type_builder.serialization_support = serialization_support;
+  nested_type_builder.allocator = *allocator;
+
+  ret = rosidl_dynamic_typesupport_dynamic_type_builder_init_from_description(
+    serialization_support, recurse_desc, allocator, &nested_type_builder);
+  rosidl_runtime_c__type_description__TypeDescription__destroy(recurse_desc);
+
+  if (ret != RCUTILS_RET_OK) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Could not construct nested type builder for field [%s]", field->name.data);
+    if (ret == RCUTILS_RET_OK) {
+      ret = RCUTILS_RET_ERROR;
+    }
```

Comment:
From what I'm seeing, the only real diff in factoring out this function is that you removed the line
```
rosidl_runtime_c__type_description__TypeDescription__destroy(recurse_desc);
```
right here. It seems that is was superfluous anyways.

Other than that, all the diff is just the variable name `field_offset`, `return` instead of `goto error`, and fixing the linter hinters.

Reply:
> right here. It seems that is was superfluous anyways.

Yeah, that was my understanding.  It should also be fixing it so that we actually always check the return value, which is what clang was originally complaining about.